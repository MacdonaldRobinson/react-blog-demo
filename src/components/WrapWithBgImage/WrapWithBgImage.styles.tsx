import styled from "styled-components";

export const WrapBgImage = styled.img`
  z-index: -1;
`;
export const WrapWithBgImageContent = styled.div`
  background-color: rgba(0, 0, 0, 0.5);
  backdrop-filter: blur(5px);
`;

export const WrapWithBgImageWrappper = styled.div`
  width: 100%;
  height: 100%;
  position: relative;
  opacity: 0;
  transition: all 1s;

  &.show {
    opacity: 1;
  }

  ${WrapBgImage}, ${WrapWithBgImageContent} {
    position: absolute;
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;
  }
`;
