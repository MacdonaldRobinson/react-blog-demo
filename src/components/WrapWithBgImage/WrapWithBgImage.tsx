import { useState } from "react";
import {
    WrapBgImage,
    WrapWithBgImageContent,
    WrapWithBgImageWrappper,
} from "./WrapWithBgImage.styles";
import clsx from "clsx";

export type TWrapWtthBgImage = {
    altText: string;
    children: React.ReactNode;
    bgImageUrl: string;
    onImageLoaded?: () => void;
};
const WrapWithBgImage = ({
    altText,
    children,
    bgImageUrl,
    onImageLoaded,
}: TWrapWtthBgImage) => {
    const [isImageLoaded, setImageLoaded] = useState<boolean>(false);
    const handleImageLoaded = () => {
        setImageLoaded(true);
        if (onImageLoaded) {
            onImageLoaded();
        }
    };

    return (
        <WrapWithBgImageWrappper className={clsx({ show: isImageLoaded })}>
            <WrapBgImage
                src={bgImageUrl}
                loading="lazy"
                onLoad={handleImageLoaded}
                alt={altText}
            />
            <WrapWithBgImageContent>{children}</WrapWithBgImageContent>
        </WrapWithBgImageWrappper>
    );
};

export default WrapWithBgImage;
