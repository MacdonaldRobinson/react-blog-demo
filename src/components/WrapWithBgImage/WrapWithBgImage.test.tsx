import { render } from "@testing-library/react";
import { describe, expect, it } from "vitest";
import WrapWithBgImage from "./WrapWithBgImage";

describe("WrapWithBgImage", () => {
    it("Snapshot compare", () => {
        // const { asFragment } = render(
        //     <WrapWithBgImage bgImageUrl={"http://imgul"}>test</WrapWithBgImage>
        // );
        // expect(asFragment()).toMatchSnapshot();
    });

    it("Must lazy load image", () => {
        const { container } = render(
            <WrapWithBgImage bgImageUrl={"http://imgul"} altText="title">
                test content
            </WrapWithBgImage>
        );

        const imgs = container.querySelectorAll("img");
        expect(imgs).toHaveLength(1);

        // Find the div immediately following the img
        const img = imgs[0];
        const divSibling = img.nextElementSibling;
        expect(divSibling).toBeInTheDocument();

        expect(divSibling?.tagName.toLowerCase()).toBe("div");
        expect(divSibling).toHaveTextContent("test content");
    });
});
