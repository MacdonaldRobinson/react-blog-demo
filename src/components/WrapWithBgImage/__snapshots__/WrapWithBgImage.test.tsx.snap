// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`WrapWithBgImage > Snapshot compare 1`] = `
<DocumentFragment>
  <div
    class="sc-dQkurY gYBnTt"
  >
    <img
      class="sc-VHjGu dxEIAc"
      loading="lazy"
      src="http://imgul"
    />
    <div
      class="sc-gnOvAp ArvQF"
    >
      test
    </div>
  </div>
</DocumentFragment>
`;
